org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/impl/XtextBuilder.java

```diff
@@ -85,6 +93,58 @@
 	public IResourceSetProvider getResourceSetProvider() {
 		return resourceSetProvider;
 	}
+
+	/**
+	 * 
+	 * @since 2.16
+	 */
+	@Singleton
+	@Beta
+	public static class BuilderPreferences {
+		/**
+		 * Name of a preference defining the scheduling rule strategy for the builder.
+		 */
+		public static final String PREF_SCHEDULING_RULE = "schedulingrule"; //$NON-NLS-1$
+		
+		private SchedulingOption schedulingOption;
+		
+		public BuilderPreferences () {
+			IPreferenceStore preferenceStore = new ScopedPreferenceStore(InstanceScope.INSTANCE,
+					Activator.PLUGIN_ID);
+
+			String schedulingRuleName = preferenceStore.getString(PREF_SCHEDULING_RULE);
+			if (schedulingRuleName.isEmpty()) {
+				schedulingOption = SchedulingOption.WORKSPACE;
+			} else {
+				schedulingOption = SchedulingOption.valueOf(schedulingRuleName);
+			}
+			
+			preferenceStore.addPropertyChangeListener(e -> {
+				if (PREF_SCHEDULING_RULE.equals(e.getProperty())) {
+					schedulingOption = SchedulingOption.valueOf(e.getNewValue().toString());
+				}
+			});
+		}
+	}
+	
+	@Inject
+	private BuilderPreferences preferences;
+	
+	/**
+	 * Options for scheduling rules.
+	 * 
+	 * @since 2.16
+	 */
+	public static enum SchedulingOption {
+		WORKSPACE,
```

Comment:
do we want proper javadoc for these?

Reply:
Not sure if we keep this configurable in the future. Assumption is that one scheduling rule is enough. We just don't know yet the best strategy.
Or do you mean just convert the single-line comments to Javadoc?

Reply:
yes i mean have the enum values properly documented